### PR TITLE
implement DbaApi

### PR DESCRIPTION
**Pull request by jakubvojacek**
implements changes suggested in https://github.com/staabm/phpstan-dba/pull/477

### CODE REVIEW

File: src/QueryReflection/BasePdoQueryReflector.php [hunk 1]
```diff
@@ -136,6 +136,15 @@ public function getResultType(string $queryString, int $fetchType): ?Type
         return $arrayBuilder->getArray();
     }
 
+    public function setupDbaApi(?DbaApi $dbaApi): void
```
Comment (staabm) [inline src/QueryReflection/BasePdoQueryReflector.php:139]: I think we should leave the method in the base class but implement it in the subclasses, so we don't need this ugly if-condition which couples the base class to the actual subclasses

File: src/QueryReflection/ChainedReflector.php [hunk 2]
```diff
@@ -61,4 +61,8 @@ public function getResultType(string $queryString, int $fetchType): ?Type
 
         return null;
     }
+
+    public function setupDbaApi(?DbaApi $dbaApi): void
+    {
+    }
```
Comment (staabm) [inline src/QueryReflection/ChainedReflector.php:67]: I think this should be

```
foreach ($this->reflectors as $reflector) {
  $reflector->setupDbaApi($dbaApi);
}
```

File: src/QueryReflection/DbaApi.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    /** @var self::API_* */
+    private $api;
+
+    /** @param self::API_* $api */
+    public function __construct($api)
+    {
+        $this->api = $api;
+    }
+
+    public function returnsDateTime(): bool
```
Comment (staabm) [inline src/QueryReflection/DbaApi.php:20]:
```suggestion
    public function returnsDateTimeImmutable(): bool
```

File: src/QueryReflection/LazyQueryReflector.php [hunk 4]
```diff
@@ -43,4 +43,8 @@ public function getResultType(string $queryString, int $fetchType): ?Type
 
         return $this->reflector->getResultType($queryString, $fetchType);
     }
+
+    public function setupDbaApi(?DbaApi $dbaApi): void
+    {
+    }
```
Comment (staabm) [inline src/QueryReflection/LazyQueryReflector.php:49]: store the `dbaApi` in a local property and pass it via `setupDbaApi` after reflector creation.

might be worth it to extract the reflector creation into a new private method

**Review by staabm — CHANGES_REQUESTED** [review]
this looks great, just noted a few nits.

File: src/QueryReflection/ChainedReflector.php (hunk 2 above)
Comment (jakubvojacek) [reply src/QueryReflection/ChainedReflector.php:70]: hopefully done

File: src/QueryReflection/DbaApi.php (hunk 3 above)
Comment (jakubvojacek) [reply src/QueryReflection/DbaApi.php]: hopefully done

File: src/QueryReflection/LazyQueryReflector.php (hunk 4 above)
Comment (jakubvojacek) [reply src/QueryReflection/LazyQueryReflector.php:57]: > store the dbaApi in a local property and pass it via setupDbaApi after reflector creation.

done

> might be worth it to extract the reflector creation into a new private method

wasnt exactly sure what do you mean

Comment (staabm) [reply src/QueryReflection/LazyQueryReflector.php:57]: `$this->reflector = ($this->reflectorFactory)();` can be found 2 times in this class.

I think we should extract a method

```
private function createReflector():QueryReflector {
    if (null === $this->reflector) {
            $this->reflector = ($this->reflectorFactory)();
            $this->reflector->setupDbaApi($this->dbaApi);
    }
    return $this->reflector;
}
```

and use it from this 2 call sites

Comment (jakubvojacek) [reply src/QueryReflection/LazyQueryReflector.php:61]: Sorry, I overlooked the 2nd usage :( Hopefully done

**Review by staabm — COMMENTED** [review]
looks great, thx.

could you please adjust all remaining `new QueryReflection()` spots in the codebase so we have a consistent use of `DbaApi` within this repo?

I think we only need a few new constants, like `DbaAPI::API_DOCTRINE_DBAL`, `DbaAPI::API_PDO`, `DbaAPI::API_MYSQLI`.
after that we are good to go.

File: src/QueryReflection/LazyQueryReflector.php (hunk 4 above)
Comment (staabm) [reply src/QueryReflection/LazyQueryReflector.php:61]: no worries. thx for working on it.

**Comment by staabm** [comment]
ok lets give it a try as is then. thank you

**Pull request merged**